### PR TITLE
v1.0.0 refactor and features

### PR DESCRIPTION
**Pull request by mattdodge**
Quite a bit of refactoring here. Includes these new features too:
- Group by (#17)
- Initial state (#19)
- Better persistence (#18)

### CODE REVIEW

**Comment by mattdodge** [comment]
@hansmosh take a look - specifically concerned about these changes
- [ ] Correct use of VersionProperty. Are we calling this one 1.0.0?
- [ ] Backwards compatibility - namely with group by - should we maybe add `group='null'` to the get_state function? The command has it by default
- [ ] StateChange blocks will always notify a state change on the _initial_ state signal now, since the blocks have an initial state. Will this break anything?

File: state_base_block.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        signals_to_notify = []
+        with self._safe_lock:
+            self.for_each_group(
+                self._process_group,
+                signals,
+                kwargs={"to_notify": signals_to_notify})
+        if signals_to_notify:
+            self.notify_signals(signals_to_notify)
+
+    def _process_group(self, signals, group, to_notify=list()):
```
Comment (hansmosh) [inline state_base_block.py]: You shouldn't use list() as parameter default value:
http://effbot.org/zone/default-values.htm

File: state_change_block.py [hunk 2]
```diff
@@ -33,21 +21,22 @@ class StateChange(StateChangeBase):
...
-                    self.notify_signals([out])
+    exclude = BoolProperty(default=True, title="Exclude Existing Fields")
+
+    def _process_group(self, signals, group, to_notify=list()):
+        """ Process the signals for a group.
+
+        Add any signals that should be passed through to the to_notify list
+        """
+        for signal in signals:
+            state_change = self._process_state(signal, group)
```
Comment (hansmosh) [inline state_change_block.py]: The class docstring that says "*state* changing from None to not None does not count as a state change." is no longer valid.

File: merge_state_block.py [hunk 3]
```diff
@@ -1,50 +1,37 @@
...
         for signal in signals:
             try:
                 is_state_sig = self.state_sig(signal)
-            except Exception as e:
-                is_state_sig = False
-                self._logger.error(
-                    "Failed determining state signal: {}".format(e)
-                )
+            except:
+                self._logger.exception("Failed determining state signal")
```
Comment (hansmosh) [inline merge_state_block.py]: If the state_sig expression fails, what is the value of is_state_sig? Or should this return?

Comment (mattdodge) [inline merge_state_block.py]: It shouldn't return but you're right, is_state_sig will need to be set to `False` here. I must have been living in JS land where you don't need that, good catch

File: state_change_block.py (hunk 2 above)
Comment (mattdodge) [inline state_change_block.py]: :+1: - just changed

File: merge_state_block.py [hunk 4]
```diff
@@ -1,50 +1,38 @@
...
                 is_state_sig = self.state_sig(signal)
-            except Exception as e:
+            except:
                 is_state_sig = False
-                self._logger.error(
-                    "Failed determining state signal: {}".format(e)
-                )
+                self._logger.exception("Failed determining state signal")
+
+            # 3 choices - state setter, state is true, or state is false
```
Comment (hansmosh) [inline merge_state_block.py]: I'm seeing just 2 choices.

**Pull request merged**